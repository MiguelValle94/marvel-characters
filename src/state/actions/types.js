export const FETCH_CHARACTERS_START = '[Characters] Fetch start';
export const FETCH_CHARACTERS_SUCCESS = '[Characters] Fetch success';
export const FETCH_CHARACTERS_ERROR = '[Characters] Fetch error';
export const FETCH_CHARACTER_DETAILS_START = '[Character Details] Fetch start';
export const FETCH_CHARACTER_DETAILS_SUCCESS =
  '[Character Details] Fetch success';
export const FETCH_CHARACTER_DETAILS_ERROR = '[Character Details] Fetch error';
export const UPDATE_FILTERS = '[Filters] Update';
