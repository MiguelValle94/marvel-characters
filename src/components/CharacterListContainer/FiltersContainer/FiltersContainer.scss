.filters {
  display: grid;
  grid-template: 1fr / 1fr 1fr 1fr;
  padding: 10px;

  &__order-by {
    grid-column: 3 / 3;
  }
}
