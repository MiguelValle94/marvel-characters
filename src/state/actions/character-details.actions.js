import {
  FETCH_CHARACTER_DETAILS_START,
  FETCH_CHARACTER_DETAILS_SUCCESS,
  FETCH_CHARACTER_DETAILS_ERROR
} from './types';
import {
  MARVEL_API_URL,
  CHARACTERS_ENDPOINT,
  API_KEY
} from '../../constants/config.constants';
import { CUSTOM_ERROR_MESSAGE } from '../../constants/error-messages.constants';

export function fetchCharacterDetailsStart() {
  return {
    type: FETCH_CHARACTER_DETAILS_START
  };
}

export function fetchCharacterDetailsSuccess(payload) {
  return {
    type: FETCH_CHARACTER_DETAILS_SUCCESS,
    payload
  };
}

export function fetchCharacterDetailsError(error) {
  return {
    type: FETCH_CHARACTER_DETAILS_ERROR,
    error
  };
}

export const fetchCharacterDetails = (id) => (dispatch) => {
  dispatch(fetchCharacterDetailsStart());
  return fetch(
    `${MARVEL_API_URL}${CHARACTERS_ENDPOINT}/${id}?apikey=${API_KEY}`
  )
    .then((response) => response.json())
    .then(({ data, code, status }) => {
      if (code >= 400 && code < 500) {
        throw status;
      }
      dispatch(fetchCharacterDetailsSuccess(data));
    })
    .catch((error) =>
      dispatch(
        fetchCharacterDetailsError({
          message:
            error && typeof error === 'string' ? error : CUSTOM_ERROR_MESSAGE
        })
      )
    );
};
