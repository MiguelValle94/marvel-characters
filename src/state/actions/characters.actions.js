import {
  FETCH_CHARACTERS_START,
  FETCH_CHARACTERS_SUCCESS,
  FETCH_CHARACTERS_ERROR
} from './types';
import {
  MARVEL_API_URL,
  CHARACTERS_ENDPOINT,
  API_KEY
} from '../../constants/config.constants';

export function fetchCharactersStart() {
  return {
    type: FETCH_CHARACTERS_START
  };
}

export function fetchCharactersSuccess(payload) {
  return {
    type: FETCH_CHARACTERS_SUCCESS,
    payload
  };
}

export function fetchCharactersError(error) {
  return {
    type: FETCH_CHARACTERS_ERROR,
    error
  };
}

export const fetchCharacters = (orderBy) => (dispatch) => {
  dispatch(fetchCharactersStart());
  return fetch(
    `${MARVEL_API_URL}${CHARACTERS_ENDPOINT}?apikey=${API_KEY}&orderBy=${orderBy}`
  )
    .then((response) => response.json())
    .then(({ data }) => dispatch(fetchCharactersSuccess(data)))
    .catch((error) => dispatch(fetchCharactersError(error)));
};
