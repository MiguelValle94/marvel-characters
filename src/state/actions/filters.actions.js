import { UPDATE_FILTERS } from './types';

export function updateFilters(payload) {
  return {
    type: UPDATE_FILTERS,
    payload
  };
}
