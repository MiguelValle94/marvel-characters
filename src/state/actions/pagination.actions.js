import { SELECT_LIMIT, SELECT_PAGE, UPDATE_PAGINATION } from './types';

export function selectLimit(limit) {
  return {
    type: SELECT_LIMIT,
    payload: limit
  };
}

export function selectPage(page) {
  return {
    type: SELECT_PAGE,
    payload: page
  };
}

export function updatePagination(payload) {
  return {
    type: UPDATE_PAGINATION,
    payload
  };
}
