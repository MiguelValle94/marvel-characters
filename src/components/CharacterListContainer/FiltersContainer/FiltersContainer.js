import React from 'react';
import { connect } from 'react-redux';

import { updateFilters } from '../../../state/actions/filters.actions';
import OrderSelector from './OrderSelector/OrderSelector';

import './FiltersContainer.scss';

function FiltersContainer({ dispatch, orderBy }) {
  function handleNameStartsWithChange(event) {
    const { length } = event.target.value;
    if (length === 0 || length >= 3) {
      handleFiltersChange(event);
    }
  }

  function handleFiltersChange(event) {
    const { name, value } = event.target;
    dispatch(updateFilters({ [name]: value }));
  }

  return (
    <form className='filters'>
      <div className='input-group'>
        <input
          className='form-control border-right-0 border'
          type='text'
          name='nameStartsWith'
          autoComplete='off'
          placeholder='Search name'
          onChange={handleNameStartsWithChange}
        />
        <span className='input-group-append'>
          <div className='input-group-text'>
            <i className='fa fa-search'></i>
          </div>
        </span>
      </div>
      <div className='filters__order-by'>
        <OrderSelector value={orderBy} onChange={handleFiltersChange} />
      </div>
    </form>
  );
}

const mapStateToProps = (state) => ({
  orderBy: state.filters.orderBy
});

export default connect(mapStateToProps)(FiltersContainer);
