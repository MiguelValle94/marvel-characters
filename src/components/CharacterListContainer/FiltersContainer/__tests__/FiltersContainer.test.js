import React from 'react';
import { mount } from 'enzyme';

import { mockStore } from '../../../../state/__mocks__/store';
import { updateFilters } from '../../../../state/actions/filters.actions';
import FiltersContainer from '../FiltersContainer';

let store;

const setUpMount = (initialState) => {
  store = mockStore(initialState);
  store.dispatch = jest.fn();
  const component = mount(<FiltersContainer store={store} />);
  return component;
};

describe('FiltersContainer', () => {
  const state = { filters: { orderBy: 'name' } };

  it('should render component', () => {
    const component = setUpMount(state);
    expect(component).toMatchSnapshot();
  });

  it('should update filters on orderBy change', () => {
    const name = 'orderBy';
    const value = '-name';
    const component = setUpMount(state);
    const select = component.find('OrderSelector');

    expect(store.dispatch).not.toHaveBeenCalled();
    select.props().onChange({ target: { value, name } });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      updateFilters({ [name]: value })
    );
  });

  it('should update filters on nameStartsWith change', () => {
    const name = 'nameStartsWith';
    const value = 'spi';
    const component = setUpMount(state);
    const input = component.find('input');

    expect(store.dispatch).not.toHaveBeenCalled();
    input.simulate('change', { target: { value, name } });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      updateFilters({ [name]: value })
    );
  });

  it('should update filters on nameStartsWith change if length === 0', () => {
    const name = 'nameStartsWith';
    const value = '';
    const component = setUpMount(state);
    const input = component.find('input');

    expect(store.dispatch).not.toHaveBeenCalled();
    input.simulate('change', { target: { value, name } });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      updateFilters({ [name]: value })
    );
  });

  it('should not update filters on nameStartsWith change if 3 > length > 0', () => {
    const name = 'nameStartsWith';
    const value = 'sp';
    const component = setUpMount(state);
    const input = component.find('input');

    expect(store.dispatch).not.toHaveBeenCalled();
    input.simulate('change', { target: { value, name } });

    expect(store.dispatch).not.toHaveBeenCalled();
  });
});
