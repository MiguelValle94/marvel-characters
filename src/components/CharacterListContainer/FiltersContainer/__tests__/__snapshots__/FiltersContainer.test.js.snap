// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersContainer should render component 1`] = `
<Connect(FiltersContainer)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <FiltersContainer
    dispatch={[MockFunction]}
    orderBy="name"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <form
      className="filters"
    >
      <div
        className="input-group"
      >
        <input
          autoComplete="off"
          className="form-control border-right-0 border"
          name="nameStartsWith"
          onChange={[Function]}
          placeholder="Search name"
          type="text"
        />
        <span
          className="input-group-append"
        >
          <div
            className="input-group-text"
          >
            <i
              className="fa fa-search"
            />
          </div>
        </span>
      </div>
      <div
        className="filters__order-by"
      >
        <OrderSelector
          onChange={[Function]}
          value="name"
        >
          <select
            className="form-control"
            name="orderBy"
            onChange={[Function]}
            value="name"
          >
            <option
              value="name"
            >
              Name Asc
            </option>
            <option
              value="-name"
            >
              Name Desc
            </option>
            <option
              value="modified"
            >
              Modified Asc
            </option>
            <option
              value="-modified"
            >
              Modified Desc
            </option>
          </select>
        </OrderSelector>
      </div>
    </form>
  </FiltersContainer>
</Connect(FiltersContainer)>
`;
